File: swift/Sources/Schemas/TraceMetadata.swift
import Foundation

public struct TraceMetadata: Codable, Hashable, Sendable {
    public let accountId: String
    public let baseHarnessVersionId: Nullable<String>?
    public let compatibilityProfileId: Nullable<String>?
    public let conversationId: Nullable<String>?
    public let converterVersion: Nullable<String>?
    public let correlationId: Nullable<String>?
    public let harnessId: Nullable<String>?
    public let harnessVersionId: Nullable<String>?
    public let importGenerationId: Nullable<String>?
    public let jobId: Nullable<String>?
    public let organizationId: Nullable<String>?
    public let productId: Nullable<String>?
    public let projectId: String
    public let source: String
    public let sourceInstance: Nullable<String>?
    public let tenantId: Nullable<String>?
    public let traceId: Nullable<String>?
    public let userId: Nullable<String>?
    /// Additional properties that are not explicitly defined in the schema
    public let additionalProperties: [String: JSONValue]

    public init(
        accountId: String,
        baseHarnessVersionId: Nullable<String>? = nil,
        compatibilityProfileId: Nullable<String>? = nil,
        conversationId: Nullable<String>? = nil,
        converterVersion: Nullable<String>? = nil,
        correlationId: Nullable<String>? = nil,
        harnessId: Nullable<String>? = nil,
        harnessVersionId: Nullable<String>? = nil,
        importGenerationId: Nullable<String>? = nil,
        jobId: Nullable<String>? = nil,
        organizationId: Nullable<String>? = nil,
        productId: Nullable<String>? = nil,
        projectId: String,
        source: String,
        sourceInstance: Nullable<String>? = nil,
        tenantId: Nullable<String>? = nil,
        traceId: Nullable<String>? = nil,
        userId: Nullable<String>? = nil,
        additionalProperties: [String: JSONValue] = .init()
    ) {
        self.accountId = accountId
        self.baseHarnessVersionId = baseHarnessVersionId
        self.compatibilityProfileId = compatibilityProfileId
        self.conversationId = conversationId
        self.converterVersion = converterVersion
        self.correlationId = correlationId
        self.harnessId = harnessId
        self.harnessVersionId = harnessVersionId
        self.importGenerationId = importGenerationId
        self.jobId = jobId
        self.organizationId = organizationId
        self.productId = productId
        self.projectId = projectId
        self.source = source
        self.sourceInstance = sourceInstance
        self.tenantId = tenantId
        self.traceId = traceId
        self.userId = userId
        self.additionalProperties = additionalProperties
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.accountId = try container.decode(String.self, forKey: .accountId)
        self.baseHarnessVersionId = try container.decodeNullableIfPresent(String.self, forKey: .baseHarnessVersionId)
        self.compatibilityProfileId = try container.decodeNullableIfPresent(String.self, forKey: .compatibilityProfileId)
        self.conversationId = try container.decodeNullableIfPresent(String.self, forKey: .conversationId)
        self.converterVersion = try container.decodeNullableIfPresent(String.self, forKey: .converterVersion)
        self.correlationId = try container.decodeNullableIfPresent(String.self, forKey: .correlationId)
        self.harnessId = try container.decodeNullableIfPresent(String.self, forKey: .harnessId)
        self.harnessVersionId = try container.decodeNullableIfPresent(String.self, forKey: .harnessVersionId)
        self.importGenerationId = try container.decodeNullableIfPresent(String.self, forKey: .importGenerationId)
        self.jobId = try container.decodeNullableIfPresent(String.self, forKey: .jobId)
        self.organizationId = try container.decodeNullableIfPresent(String.self, forKey: .organizationId)
        self.productId = try container.decodeNullableIfPresent(String.self, forKey: .productId)
        self.projectId = try container.decode(String.self, forKey: .projectId)
        self.source = try container.decode(String.self, forKey: .source)
        self.sourceInstance = try container.decodeNullableIfPresent(String.self, forKey: .sourceInstance)
        self.tenantId = try container.decodeNullableIfPresent(String.self, forKey: .tenantId)
        self.traceId = try container.decodeNullableIfPresent(String.self, forKey: .traceId)
        self.userId = try container.decodeNullableIfPresent(String.self, forKey: .userId)
        self.additionalProperties = try decoder.decodeAdditionalProperties(using: CodingKeys.self)
    }

    public func encode(to encoder: Encoder) throws -> Void {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try encoder.encodeAdditionalProperties(self.additionalProperties)
        try container.encode(self.accountId, forKey: .accountId)
        try container.encodeNullableIfPresent(self.baseHarnessVersionId, forKey: .baseHarnessVersionId)
        try container.encodeNullableIfPresent(self.compatibilityProfileId, forKey: .compatibilityProfileId)
        try container.encodeNullableIfPresent(self.conversationId, forKey: .conversationId)
        try container.encodeNullableIfPresent(self.converterVersion, forKey: .converterVersion)
        try container.encodeNullableIfPresent(self.correlationId, forKey: .correlationId)
        try container.encodeNullableIfPresent(self.harnessId, forKey: .harnessId)
        try container.encodeNullableIfPresent(self.harnessVersionId, forKey: .harnessVersionId)
        try container.encodeNullableIfPresent(self.importGenerationId, forKey: .importGenerationId)
        try container.encodeNullableIfPresent(self.jobId, forKey: .jobId)
        try container.encodeNullableIfPresent(self.organizationId, forKey: .organizationId)
        try container.encodeNullableIfPresent(self.productId, forKey: .productId)
        try container.encode(self.projectId, forKey: .projectId)
        try container.encode(self.source, forKey: .source)
        try container.encodeNullableIfPresent(self.sourceInstance, forKey: .sourceInstance)
        try container.encodeNullableIfPresent(self.tenantId, forKey: .tenantId)
        try container.encodeNullableIfPresent(self.traceId, forKey: .traceId)
        try container.encodeNullableIfPresent(self.userId, forKey: .userId)
    }

    /// Keys for encoding/decoding struct properties.
    enum CodingKeys: String, CodingKey, CaseIterable {
        case accountId = "account_id"
        case baseHarnessVersionId = "base_harness_version_id"
        case compatibilityProfileId = "compatibility_profile_id"
        case conversationId = "conversation_id"
        case converterVersion = "converter_version"
        case correlationId = "correlation_id"
        case harnessId = "harness_id"
        case harnessVersionId = "harness_version_id"
        case importGenerationId = "import_generation_id"
        case jobId = "job_id"
        case organizationId = "organization_id"
        case productId = "product_id"
        case projectId = "project_id"
        case source
        case sourceInstance = "source_instance"
        case tenantId = "tenant_id"
        case traceId = "trace_id"
        case userId = "user_id"
    }
}

File: rust/src/api/types/trace_metadata.rs
pub use crate::prelude::*;

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq)]
pub struct TraceMetadata {
    #[serde(default)]
    pub account_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub base_harness_version_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub compatibility_profile_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub conversation_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub converter_version: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub correlation_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub harness_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub harness_version_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub import_generation_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub job_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub organization_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub product_id: Option<String>,
    #[serde(default)]
    pub project_id: String,
    #[serde(default)]
    pub source: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_instance: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tenant_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub trace_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_id: Option<String>,
    /// Additional properties that are not part of the defined schema.
    #[serde(flatten)]
    pub extra: std::collections::HashMap<String, serde_json::Value>,
}

impl TraceMetadata {
    pub fn builder() -> TraceMetadataBuilder {
        <TraceMetadataBuilder as Default>::default()
    }
}

#[derive(Clone, PartialEq, Default, Debug)]
#[non_exhaustive]
pub struct TraceMetadataBuilder {
    account_id: Option<String>,
    base_harness_version_id: Option<String>,
    compatibility_profile_id: Option<String>,
    conversation_id: Option<String>,
    converter_version: Option<String>,
    correlation_id: Option<String>,
    harness_id: Option<String>,
    harness_version_id: Option<String>,
    import_generation_id: Option<String>,
    job_id: Option<String>,
    organization_id: Option<String>,
    product_id: Option<String>,
    project_id: Option<String>,
    source: Option<String>,
    source_instance: Option<String>,
    tenant_id: Option<String>,
    trace_id: Option<String>,
    user_id: Option<String>,
}

impl TraceMetadataBuilder {
    pub fn account_id(mut self, value: impl Into<String>) -> Self {
        self.account_id = Some(value.into());
        self
    }

    pub fn base_harness_version_id(mut self, value: impl Into<String>) -> Self {
        self.base_harness_version_id = Some(value.into());
        self
    }

    pub fn compatibility_profile_id(mut self, value: impl Into<String>) -> Self {
        self.compatibility_profile_id = Some(value.into());
        self
    }

    pub fn conversation_id(mut self, value: impl Into<String>) -> Self {
        self.conversation_id = Some(value.into());
        self
    }

    pub fn converter_version(mut self, value: impl Into<String>) -> Self {
        self.converter_version = Some(value.into());
        self
    }

    pub fn correlation_id(mut self, value: impl Into<String>) -> Self {
        self.correlation_id = Some(value.into());
        self
    }

    pub fn harness_id(mut self, value: impl Into<String>) -> Self {
        self.harness_id = Some(value.into());
        self
    }

    pub fn harness_version_id(mut self, value: impl Into<String>) -> Self {
        self.harness_version_id = Some(value.into());
        self
    }

    pub fn import_generation_id(mut self, value: impl Into<String>) -> Self {
        self.import_generation_id = Some(value.into());
        self
    }

    pub fn job_id(mut self, value: impl Into<String>) -> Self {
        self.job_id = Some(value.into());
        self
    }

    pub fn organization_id(mut self, value: impl Into<String>) -> Self {
        self.organization_id = Some(value.into());
        self
    }

    pub fn product_id(mut self, value: impl Into<String>) -> Self {
        self.product_id = Some(value.into());
        self
    }

    pub fn project_id(mut self, value: impl Into<String>) -> Self {
        self.project_id = Some(value.into());
        self
    }

    pub fn source(mut self, value: impl Into<String>) -> Self {
        self.source = Some(value.into());
        self
    }

    pub fn source_instance(mut self, value: impl Into<String>) -> Self {
        self.source_instance = Some(value.into());
        self
    }

    pub fn tenant_id(mut self, value: impl Into<String>) -> Self {
        self.tenant_id = Some(value.into());
        self
    }

    pub fn trace_id(mut self, value: impl Into<String>) -> Self {
        self.trace_id = Some(value.into());
        self
    }

    pub fn user_id(mut self, value: impl Into<String>) -> Self {
        self.user_id = Some(value.into());
        self
    }

    /// Consumes the builder and constructs a [`TraceMetadata`].
    /// This method will fail if any of the following fields are not set:
    /// - [`account_id`](TraceMetadataBuilder::account_id)
    /// - [`project_id`](TraceMetadataBuilder::project_id)
    /// - [`source`](TraceMetadataBuilder::source)
    pub fn build(self) -> Result<TraceMetadata, BuildError> {
        Ok(TraceMetadata {
            account_id: self
                .account_id
                .ok_or_else(|| BuildError::missing_field("account_id"))?,
            base_harness_version_id: self.base_harness_version_id,
            compatibility_profile_id: self.compatibility_profile_id,
            conversation_id: self.conversation_id,
            converter_version: self.converter_version,
            correlation_id: self.correlation_id,
            harness_id: self.harness_id,
            harness_version_id: self.harness_version_id,
            import_generation_id: self.import_generation_id,
            job_id: self.job_id,
            organization_id: self.organization_id,
            product_id: self.product_id,
            project_id: self
                .project_id
                .ok_or_else(|| BuildError::missing_field("project_id"))?,
            source: self
                .source
                .ok_or_else(|| BuildError::missing_field("source"))?,
            source_instance: self.source_instance,
            tenant_id: self.tenant_id,
            trace_id: self.trace_id,
            user_id: self.user_id,
            extra: Default::default(),
        })
    }
}
